### PR TITLE
MAINT: optimize: improve `optimize.curve_fit` doc and error message.

### PR DESCRIPTION
**Pull request by AtsushiSakai**
#### Reference issue

Follow up #16961 to address some remained comments.

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
Looks pretty safe to me, I see that there's already a guard for `propagate` but I can also see how this might improve the error messages a bit. No old tests were modified which is encouraging.

The CI timeout failure is unrelated I think--I tried to fix that over in gh-17831 but I'm hesitant to bump the timeouts even more there. Maybe we have to though..

Maybe an `optimize` regular wants to take a look before merging.

**Review by mdhaber — APPROVED** [review]
Thanks @AtsushiSakai. When I'm at a computer, I think I'll change the wording about `check_finite` - I think the simplest way to say it is that the default is True when `nan_policy` is not specified and `False` otherwise. That's correct, right?

Then I'll merge after the "Lint Everything" PR merges to avoid merge conflicts there.

File: scipy/optimize/_minpack_py.py
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:589]: When I suggested documenting 3D arrays, I had in mind that we should
a) make a more explicit decision about whether to support this officially and 
b) if so, document precisely what it means. Otherwise, it is not clear how it behaves. 

So let's take care of that in a separate PR. 
```suggestion
        functions with k predictors, and each element should be float
        convertible if it is an array like object.
```
Do you have a preference about whether it should be supported?

File: scipy/optimize/_minpack_py.py
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:629]:
```suggestion
        do contain nans. Default is True if `nan_policy` is not specified
        explicitly and False otherwise.
```

File: scipy/optimize/_minpack_py.py
Comment (mdhaber) [inline scipy/optimize/_minpack_py.py:631]:
```suggestion
```

**Comment by mdhaber** [comment]
@AtsushiSakai this looks good to go if you agree that the wording suggestion is correct!

File: scipy/optimize/_minpack_py.py
Comment (AtsushiSakai) [reply scipy/optimize/_minpack_py.py]: My preference  is supporting N-d array if possible. Because, in some fields which we are not familiar with, higher degree predictor might be useful. But I agree we don’t specify it is 3D here now.

**Comment by AtsushiSakai** [comment]
Sorry, I missed your comment. I agree your comments.

**Comment by mdhaber** [comment]
CI failure seems unrelated. Thanks @AtsushiSakai @tylerjereddy!

**Pull request merged**
